src/net/stream.rs

```diff
@@ -0,0 +1,157 @@
+use bytes::Bytes;
+use std::{
+    io,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio_util::sync::ReusableBoxFuture;
+
+use tokio::{
+    io::{AsyncRead, AsyncWrite, ReadBuf},
+    sync::Notify,
+};
+
+use crate::{world::World, ToSocketAddr};
+
+use super::{Segment, SocketPair};
+
+/// A simulated connection between two hosts.
+///
+/// All methods must be called from a host within a Turmoil simulation.
+pub struct Stream {
+    pub(crate) pair: SocketPair,
+    notify: Arc<Notify>,
+    read_fut: Option<ReusableBoxFuture<'static, ()>>,
+}
+
+impl Stream {
+    pub(crate) fn new(pair: SocketPair, notify: Arc<Notify>) -> Self {
+        Self {
+            pair,
+            notify,
+            read_fut: None,
+        }
+    }
+
+    /// Opens a connection to a remote host.
+    ///
+    /// `addr` is an address of the remote host. Anything which implements the
+    /// [`ToSocketAddr`] trait can be supplied as the address.
+    pub async fn connect(addr: impl ToSocketAddr) -> io::Result<Self> {
+        let ((local, wait), dst) = World::current(|world| {
+            let dst = world.lookup(addr);
+            (world.connect(dst), dst)
+        });
+
+        let peer = wait
+            .await
+            .map_err(|_| io::Error::new(io::ErrorKind::ConnectionRefused, dst.to_string()))?;
+
+        let pair = SocketPair { local, peer };
+        let notify = World::current(|world| world.current_host_mut().finish_connect(pair));
+
+        Ok(Stream::new(pair, notify))
+    }
+
+    fn poll_read_priv(
+        &mut self,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        let read_fut = match self.read_fut.as_mut() {
+            Some(fut) => fut,
+            None => {
+                // fast path if we can recv immediately
+                let maybe = Self::recv(self.pair, buf);
+                if let Some(res) = maybe {
+                    return Poll::Ready(res);
+                }
+
+                let notify = Arc::clone(&self.notify);
+                self.read_fut
+                    .get_or_insert(ReusableBoxFuture::new(
+                        async move { notify.notified().await },
+                    ))
+            }
+        };
+
+        match read_fut.poll(cx) {
```

Comment:
not sure if you know but there is a https://doc.rust-lang.org/std/task/macro.ready.html macro that might make this code easier. Its kinda the same as `?` for Result.

Reply:
I looked at the one in futures, but let me revisit as I agree avoiding some nesting would be nice in here.